[2단계 - 로또 리팩토링] 에드(김진우) 미션 제출합니다.

jinu0328 opened the pull request:
# 2단계 리팩토링 된 점

엘리의 피드백을 보고 기존 LottoFactory 방식과 인터페이스 방식의 차이를 고민해보았습니다.

```java
public class LottoFactory {
    public static Lotto createLotto(List<Integer> numbers) {
        return new Lotto(numbers);
    }

    public static Lotto createRandomLotto() {
        List<Integer> randomNumbers = RandomGenerator.generateNumbers(1, 45, 6);
        return new Lotto(randomNumbers);
    }
}

```

기존 static 메서드를 활용한 `LottoFactory`는 사용이 직관적이고, 별도의 팩토리 객체를 관리할 필요가 없었습니다.
덕분에 불필요한 객체 생성을 방지할 수 있다는 장점이 있었습니다.

원래 Lotto를 new를 통해 직접 생성할 계획이었지만,
로또의 생성 방식이 다양해질 가능성(자동/수동 등)을 고려하여 `LottoFactory`를 도입하게 되었습니다.
그러나, 기존 방식에는 몇 가지 한계가 있음을 발견했습니다.

- 새로운 로또 생성 방식이 추가될 경우, LottoFactory 자체를 수정해야 한다. 
- static 메서드를 사용하므로 테스트 시 유연성이 떨어진다. 
- 팩토리 클래스가 유틸성 메서드로 존재하기 때문에, 어디서든 Lotto를 생성할 수 있어 객체 생성의 일관성이 흐려질 가능성이 있다.

라는 단점이 존재할 것이라고 생각합니다.

위 문제를 해결하기 위해 이번에는 인터페이스 방식을 도입해보았습니다.

```java
public interface LottoFactory {
    Lotto createLotto();
    Lotto createLotto(List<Integer> numbers);
}
```

인터페이스를 두고 자동/수동에 따라 구현체를 나누니 다음과 같은 이점을 얻을 수 있었습니다.

- 새로운 로또 생성 방식이 필요할 때, 구현체만 추가하면 되므로 기존 코드를 수정할 필요가 없다.
- DI(의존성 주입)를 활용하여 필요한 서비스에서만 LottoFactory를 사용할 수 있다.
- 자동/수동 등 다양한 생성 방식이 존재할 경우, 각 팩토리를 분리하여 관리할 수 있어 유지보수성이 향상된다.

다만, 기존 방식에서는 static 메서드를 통해 객체를 따로 만들지 않고도 생성할 수 있었지만,
인터페이스 방식으로 변경하면서 팩토리 객체를 생성하고 주입해야 하는 과정이 추가되면서 코드가 다소 복잡해진 면이 있습니다.

그러나, Lotto 객체의 생성 방식이 더 다양해질 가능성을 고려하면, 확장성을 위해 인터페이스 방식이 더 적합하다고 판단했습니다.
단순히 객체 하나를 생성하는 정도라면 팩토리 없이 new를 통해 직접 생성하는 것이 적절할 수도 있지만,
이번 경우처럼 자동/수동 등 생성 방식이 나뉘고, 확장 가능성이 있다면 인터페이스를 도입하는 것이 더 나은 선택이 될 수 있다고 생각했습니다.

# 추가 질문

엘리의 피드백을 바탕으로 인터페이스 방식을 적용하면서, 같은 역할을 하는 메서드인데 매개변수 유무가 나뉘는 경우, 두 개의 메서드를 정의하는 것이 적절한지 고민이 되었습니다.

현재 LottoFactory에서는 다음과 같이 자동/수동 로또 생성을 위한 두 개의 메서드를 정의하였습니다.

```java
public interface LottoFactory {
    Lotto createLotto(); // 자동 로또 생성
    Lotto createLotto(List<Integer> numbers); // 수동 로또 생성
}
```

이렇게 하면 자동/수동 방식이 명확하게 구분되지만,
각 구현체(AutoLottoFactory, ManualLottoFactory)에서 불필요한 메서드를 구현해야 하는 문제가 발생합니다.

인터페이스에서 같은 역할을 하는 메서드인데, 매개변수 유무가 다르면 두 개의 메서드를 정의하는 것이 일반적인 방식인지?
만약 현재 방식이 적절하지 않다면, 더 좋은 설계 방식은 무엇일지 궁금합니다! 다음은 제가 생각해본 방식입니다.

- 인터페이스를 분리하여 AutoLottoFactory, ManualLottoFactory를 각각 다른 인터페이스로 나누는 방법
- createLotto(List<Integer> numbers) 하나만 두고, 매개변수로 null을 받으면 자동 로또로 처리하는 방법

혹시 이와 관련하여 더 좋은 설계 방식이 있을까요?

jinu0328 committed:
refactor(factory): LottoFactory를 인터페이스 방식으로 변경하여 확장성 개선

- 기존 static 메서드를 사용하는 LottoFactory를 인터페이스 방식으로 변경
- AutoLottoFactory, ManualLottoFactory를 구현하여 자동/수동 로또 생성 분리
- DI를 활용하여 서비스 레이어에서 LottoFactory를 주입받도록 수정

jinu0328 committed:
refactor(model): 메서드 분리를 통해 depth 감소

yebink commented:
안녕하세요 에드!

우선 피드백 잘 반영해주셨습니다 👍 

로또 생성 자동/수동 여부에 따라 파라미터가 달라지는데, 이 부분을 하나로 통일하기 어려워서 고민이시군요 🙂 

LottoFactory 에서 상태를 가질 수 있다면 어떨까요?
현재 에드의 코드를 보면 LottoFactory가 있고, LottoFactory 를 멤버변수로 들고있는 Service 들이 있습니다. 
그리고 그 Service 는 애플리케이션 내에서 1개만 존재해요. (나중에 학습하시겠지만 보통 Service들이 그렇습니다.)
그래서 LottoFactory도 1개만 존재해야 하기 때문에 상태를 가지기 어려워집니다. (Factory라는 네이밍도 한 몫 하는 것 같아요 ㅎㅎ)
**Lotto 객체를 생성하는 것이 목적**이라면 Lotto 내부에서 static 팩토리 메서드를 사용하는 방법도 괜찮은 것 같습니다.
```java
public class Lotto {
    private List<Integer> numbers;
    // 생략

    private Lotto(List<Integer> numbers) {
        validateNumbers(numbers);
        this.numbers = numbers;
    }
    
    public static Lotto ofRandom() {
        return new Lotto(RandomGenerator.generateNumbers(1, 45, 6));
    }
    
    public static Lotto of(List<Integer> numbers) {
        return new Lotto(numbers);
    }
    // 생략

```
이렇게요!
하지만 지금 에드의 방식도 충분히 좋습니다! 👏 



이번 미션에서 추가로 고민해보면 좋을만한 부분이 있어 피드백 드려요!

- `Container` 를 사용하신 목적이 무엇인가요?
- `Container` 의 `inputView`, `outputView` 는 미사용으로 보이네요.
- `InputConverter` 가 util 패키지에 있는데, 더 적절한 패키지가 있을까요? java.util 을 비롯해 일반적으로 util 패키지에 포함되는 클래스엔 어떤 것이 있는지 학습해보아요 🙂
- 미사용 import 들도 제거 부탁드려요.
- `PrizeTier` 의 멤버 변수에 final 을 붙이면 어떤 장점이 있을까요?
- `WinningLotto` 의 주석도 제거해볼까요?
- `PrizeTier` 가 `Lotto` 의 멤버 변수인 이유는 무엇인가요? `WinningLotto`의 값에 따라 `prizeTier` 도 변경되는데, 이는 SRP 위반이 아닐까요? `Lotto`가 `prizeTier` 를 가짐으로서 얻게 되는 장점이 무엇인가요~?
- 또, `Lotto`의 `prizeTier` 가 초기화되기 전에 `extractPrize` 메서드가 호출되면 NPE 가 발생할 것 같네요!


마지막으로, 에드가 구현한 로또 게임에서 Service 의 역할은 무엇인가요?
- `Statistics` 의 역할을 `StatisticsService` 에서 하고 있지는 않나요? 🤔 
- Service 없이 `Lotto`, `Statistics` 와 같은 모델 만으로 로또 게임을 구현할 수 있을까요?
- 지금 단계에서는 Service 를 사용해보는 것도 좋지만, 상태와 기능을 한 곳에서 정의하는 **객체지향적 프로그래밍**을 더 연습해보는 것을 추천드립니다. Service는 스프링을 하면서 계속 학습하게 되실거에요! ㅎㅎ

yebink reviewed (changes requested):
안녕하세요 에드~

2단계 미션 진행하느라 고생하셨습니다!
추가 피드백 드렸으니 확인해주세요. 😄

jinu0328 committed:
refactor(container): 사용되지 않는 멤버 제거

-InputView, OutputView 멤버 제거

jinu0328 committed:
refactor: InputConverter 패키지를 util에서 converter로 이동

jinu0328 committed:
chore: 미사용 import 제거

jinu0328 committed:
refactor(model): validatePrizeTier 메서드 구현

jinu0328 commented:
안녕하세요 엘리! 피드백 반영이 늦어져서 죄송합니다 😢 
첫번째 미션부터 엘리의 리뷰 덕분에 많이 생각해보게 되고 배운 것 같아 정말 감사합니다! 리뷰 주신 부분들에 대해 제 생각을 정리해보았습니다.

> * `Container` 를 사용하신 목적이 무엇인가요?

Container를 사용한 목적은 Controller의 생성자를 단순화하고, 객체 생성을 한 곳에서 관리하기 위해서입니다.

LottoController에서 ViewFacade와 여러 Service 객체들을 직접 생성자로 받는 방식보다는,
Container를 통해 관리하면 객체 생성 책임을 한 곳으로 집중시킬 수 있고,
LottoController의 생성자에서 필요한 파라미터 개수를 줄일 수 있다고 생각했습니다.

또한, 단순히 객체를 제공하는 역할뿐만 아니라,
서비스 객체들을 한 번만 생성하여 싱글톤처럼 재사용할 수 있도록 설계하려는 의도도 있었습니다.

다만, 다시 생각해보니 현재 Container가 스프링 컨테이너처럼 의존성 관리(DI)를 수행하는 것이 아니라,
단순히 객체를 생성하고 반환하는 역할만 하고 있어서,
굳이 필요한지 고민이 되는 부분도 있습니다.

이런 방식의 의존성 관리가 적절한 방식일까요?

> * `Container` 의 `inputView`, `outputView` 는 미사용으로 보이네요.

멤버에서 제거 완료하였습니다!

> * `InputConverter` 가 util 패키지에 있는데, 더 적절한 패키지가 있을까요? java.util 을 비롯해 일반적으로 util 패키지에 포함되는 클래스엔 어떤 것이 있는지 학습해보아요 🙂

InputConverter를 util 패키지에 두었던 이유는,
입력값(String)을 변환하는 단순한 기능을 제공하는 유틸리티 클래스라고 생각했기 때문입니다.

다만 다시 보니, 단순한 변환뿐만 아니라, 입력값의 유효성 검증까지 포함하고 있어서
꼭 util 패키지에 위치해야 하는지는 고민이 되는 부분이네요. 

엘리의 피드백을 보고 학습해보니 일반적으로 util 패키지는 순수한 유틸리티 기능(문자열 처리, 날짜 변환 등)만을 제공하는 경우가 많고,
도메인과 관련된 로직을 포함하는 경우에는 더 적절한 패키지를 찾는 것이 좋다고 생각합니다.

"입력값을 변환하고 검증하는 역할"을 고려했을 때, converter라는 별도의 패키지를 두는 것으로 리팩토링 하였습니다.

> * 미사용 import 들도 제거 부탁드려요.
> * `WinningLotto` 의 주석도 제거해볼까요?


제거 완료하였습니다!

> * `PrizeTier` 의 멤버 변수에 final 을 붙이면 어떤 장점이 있을까요?

기존에는 enum 자체가 불변이므로 내부 값도 자연스럽게 불변할 거라고 생각했는데,
멤버 변수에 final이 없으면 값이 변경될 수 있다는 점을 캐치하지 못했던 것 같습니다.

PrizeTier의 멤버 변수에 final을 추가하면 불변성(immutability)이 보장되어 값이 변경될 위험이 사라지고,
멀티스레드 환경에서도 안전하게 동작할 수 있을 것 같습니다.
또한, enum은 변하지 않는 값들을 표현하는 것이 목적인데,
final을 추가하지 않으면 멤버 변수가 의도치 않게 변경될 가능성이 있어서 enum의 원래 목적과 맞지 않을 수도 있을 것 같습니다.

피드백 반영하여 final을 추가하는 것으로 리팩토링 했습니다!

> * 또, `Lotto`의 `prizeTier` 가 초기화되기 전에 `extractPrize` 메서드가 호출되면 NPE 가 발생할 것 같네요!

예외처리 추가하였습니다!

> * `PrizeTier` 가 `Lotto` 의 멤버 변수인 이유는 무엇인가요? `WinningLotto`의 값에 따라 `prizeTier` 도 변경되는데, 이는 SRP 위반이 아닐까요? `Lotto`가 `prizeTier` 를 가짐으로서 얻게 되는 장점이 무엇인가요~?

기존에는 Lotto가 자신의 당첨 등수를 직접 가지는 것이 더 직관적이라고 생각했습니다.
"로또가 내 번호를 알고 있으니, 등수(PrizeTier)도 스스로 관리하는 것이 자연스럽지 않을까?" 라고 판단했습니다.

실제로 구현 결과 다음과 같은 장점을 가질 수 있었습니다.
1. 로또 객체 단위로 당첨 등급을 관리할 수 있어, 조회가 편리함
2. 로또 등수를 판별하는 메서드(isTierMatched())가 더 간결해짐
    (Lotto 객체가 자신의 당첨 등수를 알고 있기 때문에, PrizeTier 확인이 쉬움.)

하지만 리뷰를 받고 보니, 당첨 번호만 관리하면 되는 Lotto 객체가 자신의 등수를 판별하는 책임까지 지니게 되었다는 생각이 듭니다. 
또한 엘리의 피드백대로 현재 구조에서는 WinningLotto에 따라 prizeTier가 변경될 수 있기 때문에
결과적으로 Lotto 객체가 자신의 상태를 스스로 결정할 수 없다는 점이 구조적으로 어색하게 느껴집니다.

따라서, Lotto는 오직 "로또 번호 관리"만 담당하고, 당첨 결과는 별도의 객체(LottoResult)에서 관리하는 방식이 더 적절할 것 같습니다.

피드백 반영하여 수정해보겠습니다!
 
> 마지막으로, 에드가 구현한 로또 게임에서 Service 의 역할은 무엇인가요?
> * `Statistics` 의 역할을 `StatisticsService` 에서 하고 있지는 않나요? 🤔
> * Service 없이 `Lotto`, `Statistics` 와 같은 모델 만으로 로또 게임을 구현할 수 있을까요?
> * 지금 단계에서는 Service 를 사용해보는 것도 좋지만, 상태와 기능을 한 곳에서 정의하는 **객체지향적 프로그래밍**을 더 연습해보는 것을 추천드립니다. Service는 스프링을 하면서 계속 학습하게 되실거에요! ㅎㅎ

이번 미션에서는 비즈니스 로직을 도메인 객체(Lotto, Statistics)가 담당하고,
Service는 "비즈니스 로직을 조합하여 수행하는 역할"을 하도록 설계했습니다.

StatisticsService의 경우, 당첨 통계를 생성하는 역할을 하지만,
실제 데이터(PrizeTier, 당첨 개수, 수익률 계산 등)는 Statistics가 직접 관리합니다.

즉, 현재 Service는 도메인 객체들이 가진 기능을 조합하고, 실행 순서를 결정하는 역할을 수행하는 구조입니다.

하지만 리뷰를 받고 다시 보니, StatisticsService가 단순히 Statistics의 역할을 대신하고 있을 수도 있겠다는 점을 고민해보게 되었습니다.

"이 로직이 반드시 Service에 있어야 하는가?"
"이 역할을 도메인 객체가 할 수는 없는가?"

이런 고민을 더 해보면서 개선할 수 있는 부분을 찾아보겠습니다! 감사합니다!

jinu0328 committed:
refactor: enum 멤버 final로 선언

jinu0328 committed:
refactor: Lotto 클래스에서 PrizeTier 멤버 제거

- 당첨 번호 관리 책임에 집중할 수 있도록 prizeTier 멤버 제거
- 가격 상수 정의

jinu0328 committed:
refactor: 당첨 결과를 책임지는 LottoResult 클래스 구현

- Lotto와 해당 Lotto 객체에 해당하는PrizeTier를 계산하고 저장하는 클래스

jinu0328 committed:
refactor: LottoResult의 일급 컬렉션 LottoResults 구현

- 기존 Lottos에서 수행하던 총 당첨금 계산, Lotto 객체들을 순회하며 PrizeTier 할당 등의 역할을 LottoResults가 대신 수행
- Lottos에는 LottoResults를 생성하는 책임 부여

jinu0328 committed:
refactor: WinningLotto -> WinningNumbers 변경

- 당첨된 로또가 아닌 당첨 번호라는 의미를 전달하기 위해 클래스 이름 변경

jinu0328 committed:
chore: PrizeTier의 사용되지 않는 멤버 isBonusMatched 제거

jinu0328 committed:
refactor: 리팩토링 된 클래스를 이용하도록 controller, service 로직 수정

yebink reviewed (approved):
에드 안녕하세요!

피드백 잘 반영해주신 것 확인했습니다.
로또 미션 구현하느라 고생 많으셨고, 출석 미션도 화이팅하세요! 👍
